### DISCUSSION TITLE
Need help deciding approach for virus scanning

### DISCUSSION THREAD

**Discussion by sunilke** [drakkan/sftpgo#778] [General]
We are planning to use sftpgo for sftp server on our testing and production environment/s in coming months and will be using any of one cloud storage provider.One major concern for us is virus scanning of files being uploaded and we are enforced to use some specific virus scanning platform due to some internal reasons.

I need some help deciding which approach would be best from following:

1) As all the uploads go via local temporary directory so easiest way would be just install that virus scanning platform in same location of sftpgo deployment(container,etc) and enable auto scanning(or can scan on demand whenever user upload begins). Virus scanner will then automatically scan all the directories(that will include sftpgo's user home dirs too) so during user upload that will be going on if virus is found at any point of time in unliked file of local temp directory then it will be detected, the virus scanner then should delete that temp file and upload will fail.

2) Another approach can be instead of installing virus scanning platform , what we can do is whenever user upload is requested we let it upload on local home dir and then use api(sync flow) of virus scanning platform to scan this file and upon successful result we will move file from local home dir to cloud storage provider.In this flow ,upload to s3 will not begin until file is scanned .

These are my half baked ideas ,according to you what should be most suitable approach to support virus scanning with sftpgo. Additionally ,can you foresee any issues with these?

Are there any other sftpgo users already using some best approach/s to support virus scanning? 
Also please give your suggestions on this . Thanks

**Comment by readl1** [comment]
We are using a product call bucketav which uses a sqs queue and event notifications to scan objects uploaded to s3. Its built on ClaimAV. 

https://bucketav.com/

**Reply by sunilke** [reply]
I was looking for some virus scan on run approaches where once user requests the upload, virus scan should start and upon successful scan result it should be uploaded to required vfs. Using clamav(not the bucketav) in this scenario will be ideal but as I said we are bound to use some specific virus scanner. Anyways , thanks a lot for sharing your solution.

Once any suitable approach is finalized and I am able to accomplish this totally configurable(i.e easy to enable/disable virus scanning for user uploads/downloads) virus scanning on run , will contribute it to this project where user can use any preferred virus scanner to scan user uploads and downloads(if needed).

**Comment by JonBendtsen** [comment]
What we plan to do is use a podman pod with 2+ SFTPGo and ClamAV containers. One SFTPGo receives the file and a script is started by the event manager on each upload. This script moves the file from upload, scans it with ClamAV, and if not infected it moves the file to a directory that the 2. SFTPGo server shares. All is done in the same Volume so moving files are quick.

1. we only scan on upload because we consume files very soon after upload.
2. if you store your files for longer it might be an idea to ALSO scan them regularly.

**Comment by readl1** [comment]
We have been using bucketav (https://bucketav.com/) which works off of sns
notifications put into SQS

**Comment by drakkan** [comment]
We could also add support for real-time virus scanning directly into SFTPGo. This is quite different from the approaches you are currently using because the scan happens while the upload is in progress, so it works with any storage backend and doesn't require any additional licenses (e.g. the one for bucketav). Unfortunately no company seems interested in supporting this work

**Reply by AntonyKing7D** [reply]
Do you know how much work this would involve?

**Comment by drakkan** [comment]
> Do you know how much work this would involve?

We have not yet evaluated the effort and the real feasibility of this development, it is just an idea.
Let us know if you have a budget to support at least a part of the development. 
For paid developments you can contact us via email: support at sftpgo dot com. Thank you

**Comment by steven-vuong** [comment]
Can this be solved with an upload (or better yet a pre-upload) event? Or maybe a Custom Action hook?
When one of these events trigger, we can call an iCAP service?
